plugin/shared/grpc.go

```diff
@@ -0,0 +1,292 @@
+	user, err := m.client.AddUserWithFederatedID(context.Background(), &proto.AddUserWithFederatedIDRequest{
+		Provider: provider,
+		Id:       id,
+		Role:     role,
+	})
+	return user, err
+}
+
+func (m *GRPCClient) AttachDeviceToken(id, token string) error {
+	_, err := m.client.AttachDeviceToken(context.Background(), &proto.AttachDeviceTokenRequest{
+		Id:    id,
+		Token: token,
+	})
+	return err
+}
+
+func (m *GRPCClient) DetachDeviceToken(token string) error {
+	_, err := m.client.DetachDeviceToken(context.Background(), &proto.DetachDeviceTokenRequest{
+		Token: token,
+	})
+	return err
+}
+
+func (m *GRPCClient) Close() {
+	m.client.Close(context.Background(), new(proto.Empty))
+}
+
+func (m *GRPCClient) DeleteUser(id string) error {
+	_, err := m.client.DeleteUser(context.Background(), &proto.DeleteUserRequest{
+		Id: id,
+	})
+	return err
+}
+
+func (m *GRPCClient) FetchUsers(search string, skip, limit int) ([]*proto.User, int, error) {
+	res, err := m.client.FetchUsers(context.Background(), &proto.FetchUsersRequest{
+		Search: search,
+		Skip:   uint32(skip),
+		Limit:  uint32(limit),
+	})
+	if err != nil {
+		return nil, 0, err
+	}
+	return res.Users, int(res.Total), err
+}
+
+func (m *GRPCClient) IDByName(name string) (string, error) {
+	res, err := m.client.IDByName(context.Background(), &proto.IDByNameRequest{
+		Name: name,
+	})
+	if err != nil {
+		return "", err
+	}
+	return res.Id, err
+}
+
+func (m *GRPCClient) ImportJSON(data []byte) error {
+	_, err := m.client.ImportJSON(context.Background(), &proto.ImportJSONRequest{
+		Data: data,
+	})
+	return err
+}
+
+func (m *GRPCClient) RequestScopes(userID string, scopes []string) ([]string, error) {
+	res, err := m.client.RequestScopes(context.Background(), &proto.RequestScopesRequest{
+		UserId: userID,
+		Scopes: scopes,
+	})
+	if err != nil {
+		return nil, err
+	}
+	return res.Scopes, err
+}
+
+func (m *GRPCClient) Scopes() []string {
+	res, err := m.client.Scopes(context.Background(), &proto.Empty{})
+	if err != nil {
+		return nil
+	}
+	return res.Scopes
+}
+
+func (m *GRPCClient) ResetPassword(userID, password string) error {
+	_, err := m.client.ResetPassword(context.Background(), &proto.ResetPasswordRequest{
+		Id:       userID,
+		Password: password,
+	})
+	return err
+}
+
+func (m *GRPCClient) UpdateLoginMetadata(userID string) {
+	m.client.UpdateLoginMetadata(context.Background(), &proto.UpdateLoginMetadataRequest{
+		UserId: userID,
+	})
+}
+
+func (m *GRPCClient) UpdateUser(userID string, newUser *proto.User) (*proto.User, error) {
+	user, err := m.client.UpdateUser(context.Background(), &proto.UpdateUserRequest{
+		UserId:  userID,
+		NewUser: newUser,
+	})
+	return user, err
+}
+
+// Here is the gRPC server that GRPCClient talks to.
+type GRPCServer struct {
+	// This is the real implementation
+	Impl UserStorage
+}
+
+func (m *GRPCServer) UserByID(ctx context.Context, req *proto.UserByIDRequest) (*proto.User, error) {
```

Comment:
looks like we have both GRPC server and client in one file that does the same thing?